components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -542,70 +544,45 @@ public User updateUser(User user, Map<String, Boolean> requiredAttributes) throw
             log.debug(String.format("Listing users by filter: %s %s %s", node.getAttributeValue(), node.getOperation(),
                     node.getValue()));
         }
-        try {
-            // Apply filter enhancements for username and update expression node value.
-            applyFilterEnhancementsForUsername(node);
-
-            if ((carbonUM instanceof PaginatedUserStoreManager) && isPaginatedUserStoreAvailable()) {
-
-                // If the limit is not specified, list all the users.
-                if (limit <= 0) {
-                    userNames = getAllFilteredUsers(node, limit, offset, domainName);
-                } else {
-                    // Extract he domain name if the domain name is embedded in the filter attribute value.
-                    domainName = resolveDomainNameInAttributeValue(domainName, node);
 
-                    // Get filtered user names.
-                    userNames = getFilteredUsers(node, offset, limit, sortBy, sortOrder, domainName);
-                }
-            } else
-                userNames = filterUsersUsingLegacyAPIs(node, limit, offset);
-        } catch (NotImplementedException e) {
-            String errorMessage = String.format("System does not support filter operator: %s", node.getOperation());
-            throw new CharonException(errorMessage, e);
+        // Apply filter enhancements for username and update expression node value.
+        applyFilterEnhancementsForUsernameInNode(node);
+        try {
+            // Extract he domain name if the domain name is embedded in the filter attribute value.
+            domainName = resolveDomainNameInAttributeValue(domainName, node);
         } catch (BadRequestException e) {
             String errorMessage = String
                     .format("Domain parameter: %s in request does not match with the domain name in the attribute "
                             + "value: %s ", domainName, node.getValue());
             throw new CharonException(errorMessage, e);
-        } catch (CharonException e) {
-            String errorMessage = String.format("Error in filtering users by attribute name: %s, "
-                            + "attribute value: %s and filter operation: %s", node.getAttributeValue(), node.getValue(),
-                    node.getOperation());
+        }
+        try {
+            // If the limit is not specified, list all the users using old APIs since the new APIs must have a
+            // limit larger than zero.
+            if (limit <= 0) {
+                userNames = filterUsersUsingLegacyAPIs(node, limit, offset, domainName);
```

Comment:
Shall we refactor this if block?

```
if (!isUseLegacyAPIs(.....)) {
// use new apis
} else {
// use old apis
}

boolean isUseLegacyAPIs(....) {
 return limit > 0 && isPaginatedUserStoreAvailable() && (carbonUM instanceof PaginatedUserStoreManager);
}
```

WDYT?

Reply:
The logic behind this is to deviate the flow depending on the limit. If we change this the idea would be something else.

Reply:
I believe the concern here is to use legacy APIs or new APIs? The flow will actually deviate based on the `isUseLegacyAPIs` condition, not only the limit alone.

Addressing this will organize the logic better IMO, as the current code looks a bit long with `filterUsersUsingLegacyAPIs` appearing in two places.

Reply:
Fixed: f999948935ff2081b9362c4cbe0d6e8f5ee238db